iina/FFmpegController.m

```diff
@@ -372,4 +381,324 @@ + (NSDictionary *)probeVideoInfoForFile:(nonnull NSString *)file
+        // Convert each pixel component in the row.
+        for (int index = rowOffset; index < rowOffset + bytesPerRow; index += bytesPerComponent) {
+          uint16_t componentValue;
+          memcpy(&componentValue, &pFrameRGB->data[0][index], sizeof componentValue);
+          const _Float16 asFloat = (float)componentValue / USHRT_MAX;
+          memcpy(&pFrameRGB->data[0][index], &asFloat, sizeof asFloat);
+        }
+      }
+    }
+
+    // Determine the color space to use for the image.
+    switch (pFrame->color_primaries) {
+      default:
+        // If this message is logged then the situation needs to be investigated to determine the
+        // correct color space. Fall through and treat this as a SDR image.
+        WARN(@"Unexpected color primaries: %s (%d)",
+             av_color_primaries_name(pFrame->color_primaries), pFrame->color_primaries);
+      case AVCOL_PRI_UNSPECIFIED:
+      case AVCOL_PRI_BT709:
+        cgColorSpace = CGColorSpaceCreateDeviceRGB();
+        break;
+      case AVCOL_PRI_BT2020:
+        if (@available(macOS 11.0, *)) {
+          cgColorSpace = CGColorSpaceCreateWithName(kCGColorSpaceITUR_2100_PQ);
+        } else if (@available(macOS 10.15.4, *)) {
+            cgColorSpace = CGColorSpaceCreateWithName(kCGColorSpaceITUR_2020_PQ);
+        } else if (@available(macOS 10.14.6, *)) {
+            cgColorSpace = CGColorSpaceCreateWithName(kCGColorSpaceITUR_2020_PQ_EOTF);
+        } else {
+          cgColorSpace = CGColorSpaceCreateDeviceRGB();
+        }
+        break;
+      case AVCOL_PRI_SMPTE432:
+        if (@available(macOS 10.15.4, *)) {
+          cgColorSpace = CGColorSpaceCreateWithName(kCGColorSpaceDisplayP3_PQ);
+        } else if (@available(macOS 10.14.6, *)) {
+          cgColorSpace = CGColorSpaceCreateWithName(kCGColorSpaceDisplayP3_PQ_EOTF);
+        } else {
+          cgColorSpace = CGColorSpaceCreateDeviceRGB();
+        }
+    }
+    if (!cgColorSpace) {
+      ERROR(@"Cannot create color space");
+      return NULL;
+    }
+
+#if DEBUG
+    LOG(@"Selected %s color space for bitmap image",
+        CFStringGetCStringPtr(CGColorSpaceCopyName(cgColorSpace), CFStringGetSystemEncoding()));
+    LOG(@"Creating bitmap image with %d bits per component and %d bytes per pixel",
+        bitsPerComponent, bytesPerPixel);
+#endif
+
+    cgContext = CGBitmapContextCreate(pFrameRGB->data[0], pFrameRGB->width, pFrameRGB->height,
+        bitsPerComponent, pFrameRGB->width * bytesPerPixel, cgColorSpace, bitmapInfo);
+    if (!cgContext) {
+      ERROR(@"Cannot create bitmap context");
+      return NULL;
+    }
+    cgImage = CGBitmapContextCreateImage(cgContext);
+    if (!cgImage) {
+      ERROR(@"Cannot create bitmap image");
+      return NULL;
+    }
+
+    NSImage *image = [[NSImage alloc] initWithCGImage:cgImage size: NSZeroSize];
+    if (!image) {
+      ERROR(@"Cannot create image");
+    }
+    return image;
+  }
+  @finally {
```

Comment:
cute

Reply:
I thought this try-finally pattern was better than trying to cleanup in each of those error returns. If this was Swift I'd use `defer` after each allocation.